### PR TITLE
pkg: add support to consume snapshots

### PR DESCRIPTION
**Pull request by fromanirh**
Needed for https://github.com/jaypipes/ghw/issues/66 (see https://github.com/jaypipes/ghw/issues/66#issuecomment-738618640 and following)

Add support to allow `ghw` read from snapshots (created with `ghw-snapshot`) using
environment variables or programmatically.
This is meant to be used mostly in the testsuite, but can be useful for
users in general.

Added environment variables (and counterpart in golang pkgs):
- `GHW_SNAPSHOT_PATH` let users specify a snapshot
   that `ghw` will automatically consume.
- `GHW_SNAPSHOT_ROOT` let users specify the directory
   on which the snapshot should be unpacked.
- `GHW_SNAPSHOT_PRESERVE` is relevant iff `GHW_SNAPSHOT_ROOT` is given.
   Tells `ghw` that the directory is meant only to contain the given snapshot,
   thus `ghw` will *not* attempt to unpack it
   (and will go ahead silently!) unless the directory is empty.
- `GHW_SNAPSHOT_KEEP` if set, `ghw` will *not*
   clean up the unpacked snapshot once done, leaving it into the system.

Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

**Comment by fromanirh** [comment]
@jaypipes PRs will follow soon  against `ghw-snapshot` to capture more data from sysfs and, and after that I'll post also a PR with the initial set of snapshot blobs.

File: pkg/snapshot/unpack.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+const (
+	TargetRoot = "ghw-snapshot-*"
+)
+
+const (
+	OwnTargetDirectory = 1 << iota
+)
+
+func Cleanup(targetRoot string) error {
```
Comment (jaypipes) [inline pkg/snapshot/unpack.go:28]: docstrings on all exported functions please :)

**Review by jaypipes — COMMENTED** [review]
@fromanirh you rock brother! :)

I left some nits/suggestions inline but this looks awesome. Really appreciate your effort here!

Best,
-jay

**Comment by fromanirh** [comment]
Thanks for the quick review @jaypipes ! I'm happy you like the direction. I'll be more than happy to address your comments and fix the docs in the next uploads.

File: README.md [hunk 2]
```diff
@@ -81,6 +81,52 @@ Alternately, you can use the `ghw.WithChroot()` function like so:
...
+	"/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz",
+	"/my/safe/directory")
+)
+
+cpu, err := ghw.CPU(
+	ghw.WithSnapshotAndTargetRoot(
+		"/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz",
+		"/my/safe/directory")
+	),
+	ghw.WithSnapshotExclusive(true)
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply README.md:126]: done

File: README.md [hunk 3]
```diff
@@ -81,6 +81,52 @@ Alternately, you can use the `ghw.WithChroot()` function like so:
...
+
+cpu, err := ghw.CPU(ghw.WithSnapshotAndTargetRoot(
+	"/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz",
+	"/my/safe/directory")
+)
+
+cpu, err := ghw.CPU(
+	ghw.WithSnapshotAndTargetRoot(
+		"/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz",
+		"/my/safe/directory")
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply README.md:124]: done,

File: README.md [hunk 4]
```diff
@@ -81,6 +81,52 @@ Alternately, you can use the `ghw.WithChroot()` function like so:
...
+Set `GHW_SNAPSHOT_KEEP` to any value to enable it. If set, `ghw` will *not*
+clean up the unpacked snapshot once done, leaving it into the system.
+
+```go
+cpu, err := ghw.CPU(ghw.WithSnapshot("/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz"))
+
+
+cpu, err := ghw.CPU(ghw.WithSnapshotAndTargetRoot(
+	"/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz",
+	"/my/safe/directory")
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply README.md:118]: done

File: README.md [hunk 5]
```diff
@@ -81,6 +81,52 @@ Alternately, you can use the `ghw.WithChroot()` function like so:
...
+
+`GHW_SNAPSHOT_EXCLUSIVE` is relevant iff `GHW_SNAPSHOT_ROOT` is given.
+Set it to any value to toggle it on. This tells `ghw` that the directory is meant
+only to contain the given snapshot, thus `ghw` will *not* attempt to unpack it
+(and will go ahead silently!) unless the directory is empty.
+You can use both `GHW_SNAPSHOT_ROOT` and `GHW_SNAPSHOT_EXCLUSIVE` to make sure
+`ghw` unpacks the snapshot only once regardless of how many `ghw` packages
+(e.g. cpu, memory) access it.
+
+Set `GHW_SNAPSHOT_KEEP` to any value to enable it. If set, `ghw` will *not*
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply README.md:109]: good idea. Done!

File: pkg/option/option.go [hunk 6]
```diff
@@ -23,6 +27,34 @@ func EnvOrDefaultChroot() string {
 	return defaultChroot
 }
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/option/option.go:29]: added to all in `options.go`

File: pkg/context/context.go [hunk 7]
```diff
@@ -8,27 +8,83 @@ package context
...
 	return &Context{
-		Chroot: chrootVal,
+		Chroot:            chrootVal,
+		SnapshotPath:      snapPathVal,
+		SnapshotRoot:      snapRootVal,
+		SnapshotExclusive: snapExclusiveVal,
 	}
 }
+
+func (ctx *Context) Do(fn func() error) error {
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/context/context.go:49]: sure thing :)

File: pkg/option/option.go [hunk 8]
```diff
@@ -35,23 +67,85 @@ type Option struct {
...
+		SnapshotPath: &name,
+		SnapshotRoot: &root,
+	}
+}
+
+func WithSnapshotExclusive(value bool) *Option {
+	return &Option{
+		SnapshotExclusive: &value,
+	}
+}
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/option/option.go:111]: This looks much neater than my approach, thanks! Implementing.

File: pkg/snapshot/unpack.go (hunk 1 above)
Comment (fromanirh) [reply pkg/snapshot/unpack.go:28]: sure thing!

**Comment by fromanirh** [comment]
update: I'd love to add a couple unit test before to the next upload to address review comments. This will take a little while more.

**Comment by fromanirh** [comment]
(clicked wrong button again)

**Comment by fromanirh** [comment]
@jaypipes hey, just FYI I'm intentionally keeping the commits separated to make review (hopefully) easier, if you want me to squash me no problem at all, just let me know and I'll do that right away.

File: README.md [hunk 9]
```diff
@@ -81,6 +81,54 @@ Alternately, you can use the `ghw.WithChroot()` function like so:
...
+directory managed by `ghw`, and cleaned up when no longer needed, avoiding
+leftovers.
+
+The rest of the environment variables are relevant iff `GHW_SNAPSHOT_PATH` is given.
+`GHW_SNAPSHOT_ROOT` let users specify the directory
+on which the snapshot should be unpacked. This moves the ownership of that
+directory from `ghw` to users. `ghw` will still clean up the unpacked snapshot
+when no longer needed.
+
+`GHW_SNAPSHOT_PRESERVE` is relevant iff `GHW_SNAPSHOT_ROOT` is given.
```
Comment (jaypipes) [inline README.md:101]: Whoops :) You replaced the `GHW_SNAPSHOT_EXCLUSIVE` with `GHW_SNAPSHOT_PRESERVE` :) It was actually `GHW_SNAPSHOT_KEEP` that I was suggesting to change to `GHW_SNAPSHOT_PRESERVE`.

File: README.md [hunk 10]
```diff
@@ -81,6 +81,54 @@ Alternately, you can use the `ghw.WithChroot()` function like so:
...
+cpu, err := ghw.CPU(ghw.WithSnapshot(ghw.SnapshotOptions{
+	Path: "/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz",
+	Root: &myRoot,
+}))
+
+myOtherRoot := "/my/other/safe/directory"
+cpu, err := ghw.CPU(ghw.WithSnapshot(ghw.SnapshotOptions{
+	Path: "/path/to/linux-amd64-d4771ed3300339bc75f856be09fc6430.tar.gz",
+	Root: &myOtherRoot,
+	Preserve: true,
```
Comment (jaypipes) [inline README.md:128]: should be `Exclusive: true` :)

File: pkg/context/context.go [hunk 11]
```diff
@@ -8,27 +8,99 @@ package context
...
 )
 
 // Concrete merged set of configuration switches that act as an execution
 // context when calling internal discovery methods
 type Context struct {
-	Chroot string
+	Chroot           string
+	SnapshotPath     string
+	SnapshotRoot     string
+	SnapshotPreserve bool
```
Comment (jaypipes) [inline pkg/context/context.go:20]: SnapshotExclusive

**Review by jaypipes — COMMENTED** [review]
Hi again @fromanirh, sorry I wasn't clear in my previous review! It was the GHW_SNAPSHOT_KEEP that I was suggesting to change to GHW_SNAPSHOT_PRESERVE :)

Also, recommend git rebasing your patches in this PR into a single commit if you don't mind?

Best,
-jay

**Comment by fromanirh** [comment]
> Hi again @fromanirh, sorry I wasn't clear in my previous review! It was the GHW_SNAPSHOT_KEEP that I was suggesting to change to GHW_SNAPSHOT_PRESERVE :)
> 
> Also, recommend git rebasing your patches in this PR into a single commit if you don't mind?
> 
> Best,
> -jay

No worries, no problem at all. I'll fix and squash again. Thanks for your quick reviews!

File: README.md (hunk 9 above)
Comment (fromanirh) [reply README.md:101]: ah, my fault! fixed!

File: README.md (hunk 10 above)
Comment (fromanirh) [reply README.md:128]: fixed!

File: pkg/context/context.go (hunk 11 above)
Comment (fromanirh) [reply pkg/context/context.go:20]: fixed!

**Review by jaypipes — APPROVED** [review]
Fantastic work, @fromanirh, thank you so much! :)

**Comment by fromanirh** [comment]
> Fantastic work, @fromanirh, thank you so much! :)

My pleasure! thank you for the quick and insightful reviews!

**Pull request merged**